### DISCUSSION TITLE
Preventing SQL Injections in Custom Routes

### DISCUSSION THREAD

**Discussion by MM4MA** [pocketbase/pocketbase#5718] [Q&A]
When querying data in custom routes (`routerAdd` in pb_hooks file), I use parameterized queries like this:
```
const version = "Some user input here that could be malicious";

const record = $app.dao().findFirstRecordByFilter(
    "versions",
    "version = {:version}",
    { version: version }
);
```
I think this approach secures the code against SQL Injections when doing a query?

But what should I consider when using `$app.dao().saveRecord(record)` in a custom route (`routerAdd` in pb_hooks file) to prevent SQL Injections?

Example:
```
const record = $app.dao().findFirstRecordByFilter(
  "user",
  "id = {:id}",
  { id: userID }
);

record.set("description", "Some user input here that could be malicious");
$app.dao().saveRecord(record);
```

Does `$app.dao().saveRecord(record)` handle SQL Injection prevention automatically, or are there additional steps to take?

**Comment by ganigeorgiev** [comment]
`SaveRecord`/`DeleteRecord` are safe from sql injection because under the hood they utilizes `dbx.Params`.

What you need to be aware though is that `$app.dao().saveRecord` doesn't perform fields validation (e.g. "Nonempty/Required" checks). For that you need to use the `RecordUpsert` form as shown in https://pocketbase.io/docs/js-records/#update-record-with-data-validations.
_The above has changed with v0.23.0 and there is no longer `Dao` or forms (at least not in their current form) and `app.Save` performs validation by default (to skip validations you have to call `app.SaveNoValidate`)._
